contracts/liquidityMining/KyberFairLaunchV2.sol

```diff
@@ -0,0 +1,641 @@
+  function withdraw(uint256 _pid, uint256 _amount) external override nonReentrant {
+    _withdraw(_pid, _amount);
+  }
+
+  /**
+   * @dev Withdraw all tokens (of the sender) from pool, also harvest reward
+   * @param _pid: id of the pool
+   */
+  function withdrawAll(uint256 _pid) external override nonReentrant {
+    _withdraw(_pid, userInfo[_pid][msg.sender].amount);
+  }
+
+  /**
+   * @notice EMERGENCY USAGE ONLY, USER'S REWARDS WILL BE RESET
+   * @dev Emergency withdrawal function to allow withdraw all deposited tokens (of the sender)
+   *   and reset all rewards
+   * @param _pid: id of the pool
+   */
+  function emergencyWithdraw(uint256 _pid) external override nonReentrant {
+    PoolInfo storage pool = poolInfo[_pid];
+    UserInfo storage user = userInfo[_pid][msg.sender];
+    uint256 amount = user.amount;
+
+    user.amount = 0;
+    for (uint256 i = 0; i < rewardTokens.length; i++) {
+      UserRewardData storage rewardData = user.userRewardData[i];
+      rewardData.lastRewardPerShare = 0;
+      rewardData.unclaimedReward = 0;
+    }
+
+    pool.totalStake = pool.totalStake.sub(amount);
+
+    if (amount > 0) {
+      GeneratedToken token = pool.generatedToken;
+      if (token != GeneratedToken(0)) {
+        token.burn(msg.sender, amount);
+      }
+      IERC20Ext(pool.stakeToken).safeTransfer(msg.sender, amount);
+    }
+
+    emit EmergencyWithdraw(msg.sender, _pid, _getBlockTime(), amount);
+  }
+
+  /**
+   * @dev Harvest rewards from multiple pools for the sender
+   *   combine rewards from all pools and only transfer once to save gas
+   */
+  function harvestMultiplePools(uint256[] calldata _pids) external override {
+    require(_pids.length > 0, 'harvest: empty pool ids');
+
+    // check pids, if have save vesting duration 
+    if (_isSameVestingDuration(_pids)) {
+      address[] memory rTokens = rewardTokens;
+      uint256[] memory totalRewards = new uint256[](rTokens.length);
+      address account = msg.sender;
+      uint256 pid;
+
+      for (uint256 i = 0; i < _pids.length; i++) {
+        pid = _pids[i];
+        updatePoolRewards(pid);
+        // update user reward without harvesting
+        _updateUserReward(account, pid, false);
+
+        for (uint256 j = 0; j < rTokens.length; j++) {
+          uint256 reward = userInfo[pid][account].userRewardData[j].unclaimedReward;
+          if (reward > 0) {
+            totalRewards[j] = totalRewards[j].add(reward);
+            userInfo[pid][account].userRewardData[j].unclaimedReward = 0;
+            emit Harvest(account, pid, rTokens[j], reward, _getBlockTime());
+          }
+        }
+      }
+
+      for (uint256 i = 0; i < totalRewards.length; i++) {
+        if (totalRewards[i] > 0) {
+          _lockReward(
+            IERC20Ext(rTokens[i]),
+            account,
+            totalRewards[i],
+            poolInfo[_pids[0]].vestingDuration // use same duration
+          );
+        }
+      }
+    } 
+    else {     // else harvest one by one
```

Comment:
`} else {`
since the logic when vesting durations are not the same is much simpler, please change to:
`if (!_isSameVestingDuration(_pids)) {`
`   // do something and return`
`}`
`// do something for same vesting duration`

Reply:
yeah, handle the simple case and `return;`, then the complex case

Reply:
got it